### ISSUE TITLE
Casts and conversions for owned, boxed slices

### ISSUE THREAD

**Issue by Pr0methean** [Lokathor/bytemuck#238]
Owned slices can be cast with fewer restrictions than mutably borrowed ones, because the cast doesn't have to be reversible. This issue is a request for two features:

* `try_cast_boxed_slice`: casts `Box<[A]>` to `Box<[B]>` if they have the same layout and B is `AnyBitPattern`.
* `convert_boxed_slice`: converts `Box<[A]>` to `Box<[B]>` if they have the same layout, and takes a `Fn(A) -> B` parameter. (Could be implemented by first casting to a slice of a union type.)

An example use case for `convert_box_slice` would be to create a safe version of this function. Here, `A` is `PremultipliedColorU8` which can't be `AnyBitPattern`, but `B` is `ColorU8` which can.

https://github.com/Pr0methean/OcHd-RustBuild/blob/d64009748c6582d8934ac106b2a3954b4f6bfb98/main/src/image_tasks/png_output.rs#L314-L321

**Comment by zachs18** [comment]
Does [`bytemuck::allocation::(try_)cast_slice_box`](https://docs.rs/bytemuck/latest/bytemuck/allocation/fn.cast_slice_box.html) (requires the `extern_crate_alloc` cargo feature) do what you need for the first point?

**Comment by Pr0methean** [comment]
Yes, that looks like it would work. I'll be able to test it if and when https://github.com/RazrFalcon/tiny-skia/issues/120 is resolved.

**Comment by Pr0methean** [comment]
Confirmed that it works and requires only one `transmute` inside the loop: https://github.com/Pr0methean/OcHd-RustBuild/blob/ba81f122e912db489dfb292269caae492eff5c15/main/src/image_tasks/png_output.rs#L313-L326

**Comment by Lokathor** [comment]
So just the "convert" is still requested.

When you say that "A and B have the same layout" you mean just size/align so that the slice of target B will have the same elements as the source A?

**Comment by Pr0methean** [comment]
Yes, that's right. In fact, having the same padded size would probably be enough.
